include/ginkgo/core/solver/multigrid.hpp

```diff
@@ -0,0 +1,516 @@
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_SOLVER_MULTIGRID_HPP_
+#define GKO_PUBLIC_CORE_SOLVER_MULTIGRID_HPP_
+
+
+#include <complex>
+#include <functional>
+#include <memory>
+#include <utility>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/log/logger.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/identity.hpp>
+#include <ginkgo/core/multigrid/multigrid_level.hpp>
+#include <ginkgo/core/stop/combined.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+namespace solver {
+
+/**
+ * multigrid_cycle defines which kind of multigrid cycle can be used.
+ * It contains V, W, F, and K (KFCG/KGCR) cycle.
+ * - V, W cycle uses the algorithm according to Briggs, Henson, and McCormick: A
+ *   multigrid tutorial 2nd Edition.
+ * - F cycle uses the algorithm according to Trottenberg, Oosterlee, and
+ *   Schuller: Multigrid 1st Edition. F cycle first uses the recursive call but
+ *   second uses the V-cycle call such that F-cycle is between V and W cycle.
+ * - K(KFCG/KGCR) cycle uses the algorithm with up to 2 steps FCG/GCR from Yvan:
+ *   An aggregation-based algebraic multigrid method
+ */
+enum class multigrid_cycle { v, f, w, kfcg, kgcr };
+
+
+/**
+ * mutligrid_mid_smooth_type gives the options to handle the middle smoother
+ * behavior between the two cycles in the same level. It only affects the
+ * behavior when there's no operation between the post smoother of previous
+ * cycle and the pre smoother of next cycle. Thus, it only affects W cycle and F
+ * cycle.
+ * - origin: gives the same behavior as the original algorithm, which use posts
+ *   smoother from previous cycle and pre smoother from next cycle.
+ * - previous: only uses the post smoother of previous cycle in the mid smoother
+ * - next: only uses the pre smoother of next cycle in the mid smoother
+ * - standalone: uses the defined smoother in the mid smoother
+ */
+enum class multigrid_mid_smooth_type { origin, previous, next, standalone };
```

Comment:
I see.
I am a little wondering the pre is ambiguous between the previous and pre-smoother and post between post-smoother and the latter one (next).
By the way, do you think `multigrid_mid_behavior`  okay for this term?

Reply:
For the ambiguity, maybe you could have `pre_smooth`, `post_smooth`, `both`/`post_pre_smooth` and `standalone`.

For the name of the enum, I think you could remove "multigrid", because it's already in namespace `multigrid`, no? And having the word "smooth" in it is nice, otherwise people will be confused about what "mid" is. "Mid smoother" is not a standard term. So maybe you could use `mid_smooth_type`.

Reply:
no, this is not in the multigrid.
I can put it into the namespace, although the multigrid namespace is for multigridlevel stuff

Reply:
I use both/post_smoother/pre_smoother/standalone.